### PR TITLE
awplus: fix bug with enable password when supplied

### PR DESCRIPTION
**Pull request by sgsimpson**
## Pre-Request Checklist

- [x] Passes rubocop code analysis (try `rubocop --auto-correct`)
- [ ] Tests added or adapted (try `rake test`)
- [ ] Changes are reflected in the documentation
- [x] User-visible changes appended to [CHANGELOG.md](/CHANGELOG.md)

## Description

- #3494 introduced a bug where there is no return character when an enable password is supplied. The ssh session just hangs and times out.
- Test on `x610-5.4.5-3.7` with enable password supplied and on `x210-5.4.6-1.2` with enable = true.

### CODE REVIEW

**Comment by ytti** [comment]
cmd does include new line, of course none of the above uses of cmd would work if not. 

Perhaps more likely explanation is that prompt after enable prompt doesnt match, maybe due to ansi codes? And as a workaround you force 2nd prompt.

**Comment by ytti** [comment]
@shigechika can you look at this please

**Comment by mattie47** [comment]
👋👋👋

Just gave my old issue a read through. Any changes around this would need testing with and without enable password for both telnet and ssh. 

I don't work for Allied Telesis any more so also wouldn't have any test equipment for checking things again. 

Perhaps just note my key observations near the end of the linked issue for areas of risk

**Comment by ytti** [comment]
I wonder if we should just always send \r\n in ssh cmd as well. Isn't that what SSH clients anyhow do? We could also gate the behaviour behind variable, if to send \n or \r\n, so we could possibly set that value in model.

**Comment by mattie47** [comment]
> I wonder if we should just always send \r\n in ssh cmd as well. Isn't that what SSH clients anyhow do? We could also gate the behaviour behind variable, if to send \n or \r\n, so we could possibly set that value in model.

I'm going back 8 years in my memory here, but I have a feeling it added an extra new line in the saved config file. So if you switch between ssh and telnet, it would have a diff for an extra new line. Likely a corner case that doesn't matter, but I think I tried to avoid it.

**Comment by mattie47** [comment]
> Sorry for late response. I test on enabel=true which is without password. I will test enable with password on next week.

What model device and software version are you using?

**Comment by robertcheramy** [comment]
> I wonder if we should just always send \r\n in ssh cmd as well. Isn't that what SSH clients anyhow do? We could also gate the behaviour behind variable, if to send \n or \r\n, so we could possibly set that value in model.

I would change such a behavior on a model basis only, as the risk to break a lot of models is too high.

**Comment by mattie47** [comment]
Nice one @sgsimpson . I've added a couple PR comments. 

I think we still need to hear from @shigechika as to what was tested, and on what devices to understand why any change was needed in the first place before any merge.

**Comment by ytti** [comment]
I’m not so sure about this. If this model always wants \r\n on enable pw, making it conditional does not seem reasonable.

I was thinking of implementing newline for cfg block, which takes a string argument and is \n if not specified.

So this model would be

cfg :ssh { newline “\r\n” }

This would cause ‘cmd’ to always end at \r\n instead of bare \n.

**Comment by ytti** [comment]
https://github.com/ytti/oxidized/compare/master...feat-newline
```diff
diff --git a/lib/oxidized/input/cli.rb b/lib/oxidized/input/cli.rb
index af9a0294..d16fcf75 100644
--- a/lib/oxidized/input/cli.rb
+++ b/lib/oxidized/input/cli.rb
@@ -52,6 +52,10 @@ module Oxidized
         @password || (@password = regex)
       end

+      def newline(newline_str = "\n")
+        @newline || (@newline = newline_str)
+      end
+
       def login
         match_re = [@node.prompt]
         match_re << @username if @username
diff --git a/lib/oxidized/input/ssh.rb b/lib/oxidized/input/ssh.rb
index b94d8a75..6c3d6892 100644
--- a/lib/oxidized/input/ssh.rb
+++ b/lib/oxidized/input/ssh.rb
@@ -112,7 +112,7 @@ module Oxidized

     def cmd_shell(cmd, expect_re)
       @output = String.new('')
-      @ses.send_data cmd + "\n"
+      @ses.send_data cmd + newline
       @ses.process
       expect expect_re if expect_re
       @output
```

Can you add this, and then in the model:
```ruby
cfg :ssh do
  newline "\r\n"
end
```

And report back?

**Comment by sgsimpson** [comment]
Thanks @ytti, yes that worked perfectly. I have tested all four combinations again and each worked. I have updated this pull request to use the new newline option (once avaialble in main).

I have left the telnet password prompt fix in place as well.
